### PR TITLE
[Enhancement] Support export logs of different ranks in debug mode

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

If the `log_level` is "DEBUG", logger will export logs of different ranks like this:

1 node 8 GPUs:

```bash
work_dir/20230228_141908
├── 20230306_183634_${hostname}_device0_rank0.log
├── 20230306_183634_${hostname}_device1_rank1.log
├── 20230306_183634_${hostname}_device2_rank2.log
├── 20230306_183634_${hostname}_device3_rank3.log
├── 20230306_183634_${hostname}_device4_rank4.log
├── 20230306_183634_${hostname}_device5_rank5.log
├── 20230306_183634_${hostname}_device6_rank6.log
├── 20230306_183634_${hostname}_device7_rank7.log
└── vis_data
    ├── 20230228_141908.json
    └── scalars.json
```

2 node 2 GPUs per node

```bash
work_dirs/mmlogger/20230228_153532
├── 20230228_153532_${hostname}_device0_rank0.log
├── 20230228_153532_${hostname}_device2_rank10.log
├── 20230228_153532_${hostname}_device3_rank11.log
├── 20230228_153532_${hostname}_device4_rank12.log
├── 20230228_153532_${hostname}_device5_rank13.log
├── 20230228_153532_${hostname}_device6_rank14.log
├── 20230228_153532_${hostname}_device7_rank15.log
├── 20230228_153532_${hostname}_device1_rank1.log
├── 20230228_153532_${hostname}_device2_rank2.log
├── 20230228_153532_${hostname}_device3_rank3.log
├── 20230228_153532_${hostname}_device4_rank4.log
├── 20230228_153532_${hostname}_device5_rank5.log
├── 20230228_153532_${hostname}_device6_rank6.log
├── 20230228_153532_${hostname}_device7_rank7.log
├── 20230228_153532_${hostname}_device0_rank8.log
├── 20230228_153532_${hostname}_device1_rank9.log
└── vis_data
    └── config.py
```

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by RangiLyu — COMMENTED** [review]
Suggest adding hostname and local rank to the filename for multi-node debug.
For example:
```
import socket
        hostname = socket.gethostname()
        global_rank = get_rank()
        local_rank = int(os.environ['LOCAL_RANK'])

        self._log_dir = osp.join(
            self.work_dir, self.timestamp + '_rank' + str(global_rank) +
            '_localrank' + str(local_rank) + '_' + hostname)
```

File: mmengine/logging/logger.py [hunk 1]
```diff
@@ -168,18 +173,25 @@ def __init__(self,
...
+                if matched is None:
+                    raise ValueError(
+                        f'Invalid log file: {log_file}, please check its '
+                        'correctness')
+                filepath, filename, suffix = matched.groups()
+                hostname = _get_host_info()
+                filename = f'{filename}_{hostname}_rank{rank}'
+                if suffix:
+                    filename = f'{filename}.{suffix}'
+                log_file = osp.join(filepath, filename)
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:190]:
```suggestion
                filename, suffix = osp.splitext(osp.basename(log_file))
                hostname = _get_host_info()
                filename = f'{filename}_{hostname}_rank{rank}'
                if suffix:
                    filename = f'{filename}.{suffix}'
                log_file = osp.join(osp.dirname(log_file), filename)
```

File: mmengine/logging/logger.py [hunk 2]
```diff
@@ -168,18 +171,16 @@ def __init__(self,
...
-                    filename_list[-2] = f'{filename_list[-2]}_rank{rank}'
-                    path_split[-1] = '.'.join(filename_list)
-                else:
-                    path_split[-1] = f'{path_split[-1]}_rank{rank}'
-                log_file = os.sep.join(path_split)
+                filename, suffix = osp.splitext(osp.basename(log_file))
+                hostname = _get_host_info()
+                filename = f'{filename}_{hostname}_rank{rank}'
+                if suffix:
+                    filename = f'{filename}.{suffix}'
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:178]:
```suggestion
                filename = f'{filename}{suffix}'
```

File: mmengine/logging/logger.py [hunk 3]
```diff
@@ -162,24 +167,21 @@ def __init__(self,
...
-                    filename_list = path_split[-1].split('.')
-                    filename_list[-2] = f'{filename_list[-2]}_rank{rank}'
-                    path_split[-1] = '.'.join(filename_list)
-                else:
-                    path_split[-1] = f'{path_split[-1]}_rank{rank}'
-                log_file = os.sep.join(path_split)
+            if global_rank != 0 or log_level == 'DEBUG' or distributed:
+                filename, suffix = osp.splitext(osp.basename(log_file))
+                hostname = _get_host_info()
+                filename = f'{filename}_{hostname}_device{device_id}_' \
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:178]: if hostname is '', the filename will be `filename__device...`.

File: mmengine/logging/logger.py [hunk 4]
```diff
@@ -162,24 +167,21 @@ def __init__(self,
...
-                # rename `log_file` with rank suffix.
-                path_split = log_file.split(os.sep)
-                if '.' in path_split[-1]:
-                    filename_list = path_split[-1].split('.')
-                    filename_list[-2] = f'{filename_list[-2]}_rank{rank}'
-                    path_split[-1] = '.'.join(filename_list)
-                else:
-                    path_split[-1] = f'{path_split[-1]}_rank{rank}'
-                log_file = os.sep.join(path_split)
+            if global_rank != 0 or log_level == 'DEBUG' or distributed:
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:175]: Should `log_leval == 'DEBUG'` be changed to `logging._nameToLevel[log_level] <= logging.DEBUG`?

File: mmengine/logging/logger.py [hunk 5]
```diff
@@ -297,3 +304,38 @@ def _get_rank():
...
+        # TODO: return device id of npu and mlu.
+        if not torch.cuda.is_available():
+            return local_rank
+        num_device = torch.cuda.device_count()
+        cuda_visible_device = os.getenv('CUDA_VISIBLE_DEVICES', None)
+        if cuda_visible_device is None:
+            cuda_visible_device = list(range(num_device))
+        else:
+            cuda_visible_device = cuda_visible_device.split(',')
+        return int(cuda_visible_device[local_rank])
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:326]:
```suggestion
        cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', None)
        if cuda_visible_devices is None:
            cuda_visible_devices = list(range(num_device))
        else:
            cuda_visible_devices = cuda_visible_devices.split(',')
        return int(cuda_visible_devices[local_rank])
```

File: mmengine/logging/logger.py [hunk 6]
```diff
@@ -297,3 +313,38 @@ def _get_rank():
...
+        local_rank = int(os.getenv('LOCAL_RANK', '0'))
+        # TODO: return device id of npu and mlu.
+        if not torch.cuda.is_available():
+            return local_rank
+        num_device = torch.cuda.device_count()
+        cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', None)
+        if cuda_visible_devices is None:
+            cuda_visible_devices = list(range(num_device))
+        else:
+            cuda_visible_devices = cuda_visible_devices.split(',')
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:334]:
```suggestion
        cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', None)
        if cuda_visible_devices is None:
            num_device = torch.cuda.device_count()
            cuda_visible_devices = list(range(num_device))
        else:
            cuda_visible_devices = cuda_visible_devices.split(',')
```

File: docs/zh_cn/design/logging.md [hunk 7]
```diff
@@ -412,39 +412,42 @@ logger = MMLogger.get_instance('mmengine', log_file='tmp.log', distributed=True,
...
-    └── exp_name_rank63.log
+work_dir/20230228_141908
+├── 20230306_183634_${hostname}_device0_rank0.log
+├── 20230306_183634_${hostname}_device1_rank1.log
+├── 20230306_183634_${hostname}_device2_rank2.log
+├── 20230306_183634_${hostname}_device3_rank3.log
+├── 20230306_183634_${hostname}_device4_rank4.log
+├── 20230306_183634_${hostname}_device5_rank5.log
+├── 20230306_183634_${hostname}_device6_rank6.log
+├── 20230306_183634_${hostname}_device7_rank7.log
```
Comment (RangiLyu) [inline docs/zh_cn/design/logging.md:452]: Even if the storage is non-shared, the rank id is still the global rank. The distributed group has no relation with the storage.

File: mmengine/logging/logger.py [hunk 8]
```diff
@@ -297,3 +325,38 @@ def _get_rank():
...
+
+
+def _get_device_id():
+    """Get device id of current machine."""
+    try:
+        import torch
+    except ImportError:
+        return 0
+    else:
+        local_rank = int(os.getenv('LOCAL_RANK', '0'))
```
Comment (RangiLyu) [inline mmengine/logging/logger.py:337]: Do all distributed launchers set this env variable? Did you test torch launcher and mpi?

Comment (HAOCHENYE) [reply mmengine/logging/logger.py:337]: It seems that we've assumed that `LOCAL_RANK` has been set. 

https://github.com/open-mmlab/mmengine/blob/7a074fa4781ba10fee0305c24dc07ce90db15fff/mmengine/runner/runner.py#L864

I'm not sure will `mpi` launcher set this environment, but `torchrun`  and `_init_dist_slurm` will set this environment.

image


image

**Review by RangiLyu — APPROVED** [review]

File: mmengine/logging/logger.py [hunk 9]
```diff
@@ -162,24 +169,29 @@ def __init__(self,
...
         if log_file is not None:
-            if rank != 0:
-                # rename `log_file` with rank suffix.
-                path_split = log_file.split(os.sep)
-                if '.' in path_split[-1]:
-                    filename_list = path_split[-1].split('.')
-                    filename_list[-2] = f'{filename_list[-2]}_rank{rank}'
-                    path_split[-1] = '.'.join(filename_list)
+            world_size = _get_world_size()
+            if (global_rank != 0 or log_level == 'DEBUG'
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:180]: Should `log_level == 'DEBUG'` be replaced with `log_level <= logging.DEBUG`?

File: mmengine/logging/logger.py [hunk 10]
```diff
@@ -134,8 +138,9 @@ class MMLogger(Logger, ManagerMixin):
             If `logger_name` is not defined, defaults to 'mmengine'.
         log_file (str, optional): The log filename. If specified, a
             ``FileHandler`` will be added to the logger. Defaults to None.
-        log_level (str): The log level of the handler and logger. Defaults to
-            "NOTSET".
+        log_level (str): The log level of the handler. Defaults to
+            "INFO". If log level is "DEBUG", distributed logs will be saved
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:142]:
```suggestion
            'INFO'. If log level is "DEBUG", distributed logs will be saved
```

File: mmengine/logging/logger.py (hunk 9 above)
Comment (zhouzaida) [reply mmengine/logging/logger.py:180]:
```python
# rename `log_file` with rank suffix
if (distributed or log_level <= logging.DEBUG) and world_size > 1:
    pass
```

File: mmengine/logging/logger.py [hunk 11]
```diff
@@ -162,24 +169,29 @@ def __init__(self,
...
-                    path_split[-1] = f'{path_split[-1]}_rank{rank}'
-                log_file = os.sep.join(path_split)
+                    # Omit hostname if it is empty
+                    filename = (f'{filename}_device{device_id}_'
+                                f'rank{global_rank}{suffix}')
+                log_file = osp.join(osp.dirname(log_file), filename)
             # Save multi-ranks logs if distributed is True. The logs of rank0
             # will always be saved.
-            if rank == 0 or distributed:
+            if global_rank == 0 or distributed or log_level <= logging.DEBUG:
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:194]:
```suggestion
            if global_rank == 0 or is_distrbuted_log:
```

**Review by zhouzaida — APPROVED** [review]

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**